hw05-bytecode/src/main/java/skislitsyn/utils/Ioc.java

```diff
@@ -0,0 +1,73 @@
+package skislitsyn.utils;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.HashMap;
+import java.util.HashSet;
+
+import skislitsyn.logging.Log;
+
+public class Ioc {
+    private Ioc() {
+    }
+
+    public static Object createMyClass(Object obj) {
+	InvocationHandler handler = new MyInvocationHandler(obj);
+	return Proxy.newProxyInstance(Ioc.class.getClassLoader(), obj.getClass().getInterfaces(), handler);
+    }
+
+    static class MyInvocationHandler implements InvocationHandler {
+	private final Object myClass;
+	private final HashMap<Class<?>, HashSet<Method>> loggableInterfaces = new HashMap<>();
+
+	MyInvocationHandler(Object myClass) {
+	    this.myClass = myClass;
+	    Method[] methods = myClass.getClass().getDeclaredMethods();
+	    for (Method method : methods) {
+		if (method.isAnnotationPresent(Log.class)) {
+		    Class<?>[] interfaces = myClass.getClass().getInterfaces();
+		    for (Class<?> i : interfaces) {
+			try {
+			    Method interfaceMethod = i.getDeclaredMethod(method.getName(), method.getParameterTypes());
+			    HashSet<Method> methodsSet = null;
+			    if (loggableInterfaces.containsKey(i)) {
+				methodsSet = loggableInterfaces.get(i);
+			    } else {
+				methodsSet = new HashSet<>();
+			    }
+			    methodsSet.add(interfaceMethod);
+			    loggableInterfaces.put(i, methodsSet);
+			} catch (NoSuchMethodException e) {
+			    // Interface do not have the method with annotation Log
+			} catch (SecurityException e) {
+			    throw new RuntimeException(e.getMessage());
```

Comment:
Так не надо делать.
Надо так:
```suggestion
			    throw new RuntimeException(e);
```

А почему я спрошу у вас.
Если ответ не найдете, то позже напишу.

Вообще ловить здесь разные типы исключения не вижу смысла.
Я бы просто ловил Exception
```java
catch (Exception e)
```

То, что обернули в RuntimeException это хорошо.

Reply:
Исправил на пробрасывание объекта исключения, полагаю, соль в том, что при этом пробрасываемое наверх исключение содержит в себе полную информацию о породившем его исключении со всеми дополнительными полями (при наличии), а не только текст сообщения, который зачастую бывает скуден.

Насчёт выделения NoSuchMethodException - здесь оно нужно по логике программы, т.к. если обхект будет имплеменитровать несколько интерфейсов, мы не должны падать при  поиске нужного интерфейса.

Reply:
Да, чтобы можно было получить всю цепочку исключений и при необходимости добраться до реальной причины падения.

Пояснения по NoSuchMethodException принимаются.

Reply:
Если не смотрели, то посмотрите открытый урок по исключениям https://www.youtube.com/watch?v=nP-yT6tmbGs